### PR TITLE
Add security providers in kafka security config

### PR DESCRIPTION
**Pull request by saisandeep**
Adds a new custom security provider class configuration.
The configured classes are added to the Security provider via Security.addProvider api
KIP - https://cwiki.apache.org/confluence/display/KAFKA/KIP-492%3A+Add+java+security+providers+in+Kafka+Security+config

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+package org.apache.kafka.common.config;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.Provider;
+import java.security.Security;
+import java.util.Map;
+
+public class SecurityConfig {
```
Comment (harshach) [inline clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java:22]: Is this class for unifying common configs across ssl and sasl?

Comment (saisandeep) [reply clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java:22]: Yes, this file is expected to contains all the common config for SASL and SSL. Will add a comment for the same

**Comment by jeffhuang26** [comment]
How do we handle a scenario that some providers require more information for installing providers? 

For instance, Bouncy Castle(BC) provider requires input parameter "fips:BCFIPS" for enabling FIPS mode. 
Example:
Static Configuration in java.security file:
security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider  fips:BCFIPS
Dynamic Installation 
Security.addProvider(new BouncyCastleJsseProvider(“fips:BCFIPS”)) 

So I suggested we might consider providing more info for the new config property, example like:
security.provider.info=classname of provider/name of provider/initial parameters,
Example for BC case:
security.provider.info=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider/BC/fips:BCFIPS,sun.security.provider.Sun/SUN,
Basically info for each provider will consist of three pieces information: name of class, name of provider(for unit testing purpose),initial parameter for instantiating class.
Still use comma ","  for separating each provider info.

Also new class handling installing provider should consider any other component could use this if it run in different Java vm.

Jeff Huang,

File: clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Contains the common security config for SSL and SASL
+ */
+public class SecurityConfig {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecurityConfig.class);
+
+    public static final String SECURITY_PROVIDER_CLASS_CONFIG = "security.provider.class";
+    public static final String SECURITY_PROVIDER_CLASS_DOC = "Java Provider to register security algorithms";
+
+    public static void addConfiguredSecurityProviders(Map<String, ?> configs) {
```
Comment (harshach) [inline clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java]: adding the registraiton part into Config file doesn't look good to me. Generally speaking Config classes are only there to express the config members and validation around them but not to take any action. SecurityUtils.java would be a better place.

File: clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Map;
+
+/**
+ * Contains the common security config for SSL and SASL
+ */
+public class SecurityConfig {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SecurityConfig.class);
+
+    public static final String SECURITY_PROVIDER_CLASS_CONFIG = "security.provider.class";
+    public static final String SECURITY_PROVIDER_CLASS_DOC = "Java Provider to register security algorithms";
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java]: Is the description "Java Provider to register security algorithms" accurate? It should be something like "A list of Java providers will be installed"?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java [hunk 4]
```diff
@@ -66,6 +67,7 @@
         this.configs = Collections.unmodifiableMap(configs);
         this.protocol = (String) configs.get(SslConfigs.SSL_PROTOCOL_CONFIG);
         this.provider = (String) configs.get(SslConfigs.SSL_PROVIDER_CONFIG);
+        SecurityConfig.addConfiguredSecurityProviders(this.configs);
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslEngineBuilder.java]: Will the function "addConfiguredSecurityProviders"  be called to install Java providers at the very beginning of starting Java process before any Java classes? There is no need to call the function every class on same JVM? For instance, for Kafka, we should call the function in function "startup" in KafkaServer? 

Also we need install other providers in addition to TSL provider such as provider for SASL. We need make sure all providers need be installed before use it.

**Comment by saisandeep** [comment]
> How do we handle a scenario that some providers require more information for installing providers?
> 
> For instance, Bouncy Castle(BC) provider requires input parameter "fips:BCFIPS" for enabling FIPS mode.
> Example:
> Static Configuration in java.security file:
> security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
> Dynamic Installation
> Security.addProvider(new BouncyCastleJsseProvider(“fips:BCFIPS”))
> 
> So I suggested we might consider providing more info for the new config property, example like:
> security.provider.info=classname of provider/name of provider/initial parameters,
> Example for BC case:
> security.provider.info=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider/BC/fips:BCFIPS,sun.security.provider.Sun/SUN,
> Basically info for each provider will consist of three pieces information: name of class, name of provider(for unit testing purpose),initial parameter for instantiating class.
> Still use comma "," for separating each provider info.
> 
> Also new class handling installing provider should consider any other component could use this if it run in different Java vm.
> 
> Jeff Huang,

Hi Jeff, the BouncyCastleJsseProvider example you've provided is one case taking in an constructor parameter, apart from this, there are lots of other provider Constructors which take in complex arguments like a map or an object. Addressing just the string case will be solving for a specific use case leaving out more generic cases. The base "security.Provider" currently doesn't provide any configure methods as of now; if so, we might have evaluated sending in a configuration map. So, probably the best way to handle this with the limited functions provided by "security.Providers" class is making the provider read from environmental variables or from a properties file.

**Comment by jeffhuang26** [comment]
> > How do we handle a scenario that some providers require more information for installing providers?
> > For instance, Bouncy Castle(BC) provider requires input parameter "fips:BCFIPS" for enabling FIPS mode.
> > Example:
> > Static Configuration in java.security file:
> > security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
> > Dynamic Installation
> > Security.addProvider(new BouncyCastleJsseProvider(“fips:BCFIPS”))
> > So I suggested we might consider providing more info for the new config property, example like:
> > security.provider.info=classname of provider/name of provider/initial parameters,
> > Example for BC case:
> > security.provider.info=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider/BC/fips:BCFIPS,sun.security.provider.Sun/SUN,
> > Basically info for each provider will consist of three pieces information: name of class, name of provider(for unit testing purpose),initial parameter for instantiating class.
> > Still use comma "," for separating each provider info.
> > Also new class handling installing provider should consider any other component could use this if it run in different Java vm.
> > Jeff Huang,
> 
> Hi Jeff, the BouncyCastleJsseProvider example you've provided is one case taking in an constructor parameter, apart from this, there are lots of other provider Constructors which take in complex arguments like a map or an object. Addressing just the string case will be solving for a specific use case leaving out more generic cases. The base "security.Provider" currently doesn't provide any configure methods as of now; if so, we might have evaluated sending in a configuration map. So, probably the best way to handle this with the limited functions provided by "security.Providers" class is making the provider read from environmental variables or from a properties file.



Jeff Huang. 

> > How do we handle a scenario that some providers require more information for installing providers?
> > For instance, Bouncy Castle(BC) provider requires input parameter "fips:BCFIPS" for enabling FIPS mode.
> > Example:
> > Static Configuration in java.security file:
> > security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
> > Dynamic Installation
> > Security.addProvider(new BouncyCastleJsseProvider(“fips:BCFIPS”))
> > So I suggested we might consider providing more info for the new config property, example like:
> > security.provider.info=classname of provider/name of provider/initial parameters,
> > Example for BC case:
> > security.provider.info=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider/BC/fips:BCFIPS,sun.security.provider.Sun/SUN,
> > Basically info for each provider will consist of three pieces information: name of class, name of provider(for unit testing purpose),initial parameter for instantiating class.
> > Still use comma "," for separating each provider info.
> > Also new class handling installing provider should consider any other component could use this if it run in different Java vm.
> > Jeff Huang,
> 
> Hi Jeff, the BouncyCastleJsseProvider example you've provided is one case taking in an constructor parameter, apart from this, there are lots of other provider Constructors which take in complex arguments like a map or an object. Addressing just the string case will be solving for a specific use case leaving out more generic cases. The base "security.Provider" currently doesn't provide any configure methods as of now; if so, we might have evaluated sending in a configuration map. So, probably the best way to handle this with the limited functions provided by "security.Providers" class is making the provider read from environmental variables or from a properties file.

Hi, This is my question. 
I can statically configure BC providers in java.security with following format, JVM can load the provider with additional parameter "fips:BCFIPS" and it works fine. Why we cannot provide same configuration format in Kafka to allow specify additional parameter in addition to class name for Provider.
The following configuration is tested and working. JVM can load and handle. 
security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider  fips:BCFIPS

So I think in Kafka, we might follow same pattern like static config in java.security file, so new config property would be:
security.providers=provider1  <optional parameter>,provider2

thanks!

Jeff.

**Comment by jeffhuang26** [comment]
These are my suggestions based on discussion with Rajini @rajinisivaram and on my current project relate to FIPS using Bouncy Castle(BC) provider which is compliant and certified provider and used by many company want to do business with government.
1. The new security config would be:
"security.providers=provider1 class  <optional parameter>,provider2". Format would fellow same pattern in java.security. 
security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
2. Create a new class such as SecurityProvidersInstaller or SecurityProvidersLoader. This class will implement a default function addConfiguredSecurityProviders or we might call configure() ? without consider operational parameter.
3. Any customized special handling like BC could extend the class such as BCSecurityProvidersInstaller and overwrite the function addConfiguredSecurityProviders.
4. This class will be needed by both broker and clients such as producer, consumer, so we need put this class into appropriate package.  
5. Only install all providers, such as for TLS, SASL, once per JVM.

Thanks!

Jeff Huang.

**Comment by saisandeep** [comment]
> These are my suggestions based on discussion with Rajini @rajinisivaram and on my current project relate to FIPS using Bouncy Castle(BC) provider which is compliant and certified provider and used by many company want to do business with government.
> 
> 1. The new security config would be:
>    "security.providers=provider1 class  ,provider2". Format would fellow same pattern in java.security.
>    security.provider.1=org.bouncycastle.jsse.provider.BouncyCastleJsseProvider fips:BCFIPS
> 2. Create a new class such as SecurityProvidersInstaller or SecurityProvidersLoader. This class will implement a default function addConfiguredSecurityProviders or we might call configure() ? without consider operational parameter.
> 3. Any customized special handling like BC could extend the class such as BCSecurityProvidersInstaller and overwrite the function addConfiguredSecurityProviders.
> 4. This class will be needed by both broker and clients such as producer, consumer, so we need put this class into appropriate package.
> 5. Only install all providers, such as for TLS, SASL, once per JVM.
> 
> Thanks!
> 
> Jeff Huang.

Ok, these suggestions make sense and sound reasonable. We'll go ahead with the following new changes

1. Creation of a new SecurityProvidersLoader interface
2. This interface contains a method configureProvider(or a better name) which takes in a Map<String, ?> properties object
3. Any configuration of Providers is expected to be implemented in the configureProvider method
4. This method is passed the properties passed to the broker/producer/consumer

**Comment by jeffhuang26** [comment]
> 1. security.providers

Will the new config property be "security.providers"?

Thanks!

**Comment by saisandeep** [comment]
> > 1. security.providers
> 
> Will the new config property be "security.providers"?
> 
> Thanks!

We can go ahead with "security.provider.classes" as mentioned in the KIP, unless there is some problem with it.

**Comment by jeffhuang26** [comment]
> > > 1. security.providers
> > 
> > 
> > Will the new config property be "security.providers"?
> > Thanks!
> 
> We can go ahead with "security.provider.classes" as mentioned in the KIP, unless there is some problem with it.

Actually the main idea for this KIP will allow dynamically configure security providers. so I still think "security.providers",  from long term view and flexible aspect,  better than "security.provider.classes" to represent the main idea even both need classes name. Also using "security.providers" will be consistent with format in static configuration in java.security file.  

Thanks!

Jeff.

File: clients/src/main/java/org/apache/kafka/common/security/SecurityProviderGenerator.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.apache.kafka.common.security;
+
+import java.security.Provider;
+import java.util.Map;
+
+/**
+ * An interface for generating security providers.
+ */
+public interface SecurityProviderGenerator {
```
Comment (harshach) [inline clients/src/main/java/org/apache/kafka/common/security/SecurityProviderGenerator.java]: What is the intent behind this interface. It looks like only used for test cases?

**Comment by saisandeep** [comment]
@jeffhuang26 please review

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java [hunk 6]
```diff
@@ -34,4 +43,25 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
...
+        } catch (ClassCastException e) {
+            LOGGER.error("Generator provided through " + SecurityConfig.SECURITY_PROVIDERS_CONFIG +
+                    " are expected to be sub-classes of SecurityProviderGenerator");
+        } catch (ClassNotFoundException cnfe) {
+            LOGGER.error("Unrecognized security provider class", cnfe);
+        } catch (IllegalAccessException | InstantiationException e) {
+            LOGGER.error("Unexpected implementation of security provider class", e);
+        }
+    }
+
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: Suggestion on above function addConfiguredSecurityProviders:
1.
I think using following format for security.providers:
security.providers=provider1_classname,provider2_generator_classname key1:val1 key2:val2,...
2.
So when parsing above config, if there is no parameters following provider1_classname, then we can think provider1_classname is java Provider, then insert it; if there are key:value pair parameters following provider2_generator_classname, we can think provider2_generator_classname is SecurityProviderGenerator, then create "Map<String, ?> config" from key:value pair parameters, then call configure and getProvider of instance of SecurityProviderGenerator. This way we can handle all different scenarios in the future.

Comment (harshach) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: @jeffchao traditionally Kafka used key,value pairs in properties and pass it everywhere and each implementation takes look at this config and pulls their interested key,value pairs.
Example, authorizer interface https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/security/auth/Authorizer.scala#L35 . The pluggable class when it gets instantiated a configure method will be called and all the key,value in server.properties will be passed and it will pick whats relevant to the class.
We can do the same here instead of asking users append key,values into the a config which is hard to configure and hard to get it right.

Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: I got what you said. But for this case, key/value pairs might be very specific to the provider. 
1. Do you mean we can configure SecurityProviderGenerator like JAAS that will configure class name and key/value pairs parameters together? 
2. How do we configure a Provider which will not need key/value pairs?  Do we require a wrapper which will  implement SecurityProviderGenerator for the Provider?

Comment (saisandeep) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: 1. No, we're not going with that. You can configure any properties you need in the server.properties file as key value pairs just like other config params(not like JAAS) and all the config param will be passed to the configure method of the SecurityProviderGenerator(renamed to SecurityProviderCreator). Your implementation of SecurityProviderCreator can read only the config that makes sense to it.
2. All the providerCreator classes that don't need the configure method can skip overriding this method. A default implementation of configure method is added in SecurityProviderCreator.

cc: @harshach

Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: ok. Just curious what is target date when all implementation for this will be merged?  cc@saisandeep

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java [hunk 7]
```diff
@@ -34,4 +43,25 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
...
 
+    public static void addConfiguredSecurityProviders(Map<String, ?> configs) {
+        String securityProviderClassesStr = (String) configs.get(SecurityConfig.SECURITY_PROVIDERS_CONFIG);
+        if (securityProviderClassesStr == null || securityProviderClassesStr.equals("")) {
+            return;
+        }
+        try {
+            String[] securityProviderClasses = securityProviderClassesStr.replaceAll("\\s+", "").split(",");
+            for (int index = 0; index < securityProviderClasses.length; index++) {
+                SecurityProviderCreator securityProviderCreator = (SecurityProviderCreator) Class.forName(securityProviderClasses[index]).newInstance();
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:54]: Need call securityProviderCreator.configure(config); here?

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java (hunk 6 above)
Comment (saisandeep) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: It really depends on when the PR gets accepted. I'll try to merge this as soon as it gets accepted.

Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: It seems I also could approve it. I will read all code tomorrow and work with you to get this approved.

File: clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.apache.kafka.common.config;
+
+/**
+ * Contains the common security config for SSL and SASL
+ */
+public class SecurityConfig {
+
+    public static final String SECURITY_PROVIDERS_CONFIG = "security.providers";
+    public static final String SECURITY_PROVIDERS_DOC = "A list of generators each returning a provider " +
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/config/SecurityConfig.java]: Change "generators" to "creators" ?

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java [hunk 9]
```diff
@@ -174,6 +175,11 @@
                                         Importance.MEDIUM,
                                         CLIENT_DNS_LOOKUP_DOC)
                                 // security support
+                                .define(SecurityConfig.SECURITY_PROVIDERS_CONFIG,
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java]: Do we need follow same pattern as other properties to define a new property which refers to SECURITY_PROTOCOL_CONFIG?
Example: 
public static final String SECURITY_PROVDERS_CONFIG = SecurityConfig.SECURITY_PROVIDERS_CONFIG;

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java [hunk 10]
```diff
@@ -486,6 +487,11 @@
                                         Importance.MEDIUM,
                                         ALLOW_AUTO_CREATE_TOPICS_DOC)
                                 // security support
+                                .define(SecurityConfig.SECURITY_PROVIDERS_CONFIG,
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Same comments as SECURITY_PROVDERS_CONFIG.

File: clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java [hunk 11]
```diff
@@ -341,6 +342,11 @@
                                         CommonClientConfigs.DEFAULT_SECURITY_PROTOCOL,
                                         Importance.MEDIUM,
                                         CommonClientConfigs.SECURITY_PROTOCOL_DOC)
+                                .define(SecurityConfig.SECURITY_PROVIDERS_CONFIG,
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java]: Same comment as SECURITY_PROVDERS_CONFIG.

File: clients/src/main/java/org/apache/kafka/common/security/SecurityProviderCreator.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * An interface for generating security providers.
+ */
+public interface SecurityProviderCreator {
+
+    /**
+     * Configure method is used to configure the generator to create the Security Provider
+     * @param config configuration parameters for initialising security provider
+     */
+    default void configure(Map<String, ?> config) {
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/security/SecurityProviderCreator.java:33]: interface Configurable has same definition "void configure(Map<String, ?> configs);", need extend from Configurable or no need?

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java (hunk 6 above)
Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java:67]: .

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java [hunk 13]
```diff
@@ -34,4 +43,26 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
...
+                securityProviderCreator.configure(configs);
+                Security.insertProviderAt(securityProviderCreator.getProvider(), index + 1);
+            }
+        } catch (ClassCastException e) {
+            LOGGER.error("Generator provided through " + SecurityConfig.SECURITY_PROVIDERS_CONFIG +
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: Change Generator to Creator?

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java [hunk 14]
```diff
@@ -34,4 +43,26 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
...
+                    " are expected to be sub-classes of SecurityProviderCreator");
+        } catch (ClassNotFoundException cnfe) {
+            LOGGER.error("Unrecognized security provider class", cnfe);
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: Unrecognized security provider creator class?

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java [hunk 15]
```diff
@@ -34,4 +43,26 @@ public static KafkaPrincipal parseKafkaPrincipal(String str) {
...
+        } catch (IllegalAccessException | InstantiationException e) {
+            LOGGER.error("Unexpected implementation of security provider class", e);
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: Unexpected implementation of security provider creator class?

File: clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java [hunk 16]
```diff
@@ -90,6 +91,8 @@ public void configure(Map<String, ?> configs) throws KafkaException {
         if (clientAuthConfigOverride != null) {
             nextConfigs.put(BrokerSecurityConfigs.SSL_CLIENT_AUTH_CONFIG, clientAuthConfigOverride);
         }
+        nextConfigs.put(SecurityConfig.SECURITY_PROVIDERS_CONFIG,
+                configs.get(SecurityConfig.SECURITY_PROVIDERS_CONFIG));
```
Comment (jeffhuang26) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java]: Remove about two lines code and something like below? 
copyMapEntries(nextConfigs, configs, SslConfigs.NON_RECONFIGURABLE_CONFIGS);
copyMapEntries(nextConfigs, configs, SslConfigs.RECONFIGURABLE_CONFIGS);
copyMapEntries(nextConfigs, configs, SecurityConfig. SECURITY_PROVIDERS_CONFIG)

File: clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java (hunk 13 above)
Comment (maulin-vasavada) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: Here we rely on insertProviderAt() programatic way BUT if in the application's context somebody else calls Security.insertProviderAt(provider,1) that provider will be given the priority for any conflicting Provider services+algorithms. This code works well if you have exclusive services+algorithms example SPIFFE but if you are writing a provider for Standard algorithms example TrustManagerFactory.PKIX then you may run into trouble since your insertProviderAt() call got overridden by somebody else in the application context/startup.

When that happens I don't know easy way to fix it. I think It is important to call this out.

Comment (maulin-vasavada) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: We have this kind of challenges today in some other context. Hence speaking from the real experience! :)

Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: Set provider only at the beginning of loading application per JVM will avoid what you said.   I think current solution is fine too.  We should use one provider for JSSE per JVM?

Comment (maulin-vasavada) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: So let us say - we already have a single provider for JSSE but for Kafka we need one more to override with some of the common services+algorithm (example only override TrustManagerFactory.PKIX). We are at the mercy of the initialization sequence of calls, isn't it? Here the different providers could be owned by different teams and when we are in bigger infra setup it may be difficult to overcome this technical limitation. Only for Kafka applications the init sequence is different vs rest of the infra in the company.

Basically, we are making an assumption which may not hold true and then we will be really stuck. What I am suggesting is - If we are calling out that limitation with these changes it is okay but otherwise it will result in a bug.

Comment (jeffhuang26) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: I think, for most scenarios, using a single provider for JSSE per JVM should be fine.  At some cases, If need specific providers for some components, then the component self need explicitly specify the name of the provider such as:
public static final KeyManagerFactory getInstance(String algorithm, String provider)?

Comment (maulin-vasavada) [reply clients/src/main/java/org/apache/kafka/common/utils/SecurityUtils.java]: That has its own complications because if my provider is only providing TrustManagerFacotry.PKIX then it can't provider other services+algorithms that might be expected in calls like SSLContext.getInstance(String protocol, String provider). SSLContext.getInstance might look for TLSv1.1 or TLSv1.2 etc which my provider doesn't really have and I don't have a way to fallback anymore once I go route of specifying "provider" in getInstance() calls. 

In short - once we have a Provider providing a Standard service+algorithm we may have to implement other services+algorithm also otherwise it may not work (like I mentioned for SSLContext)

**Comment by jeffhuang26** [comment]
HI Saisandeep,

@saisandeep  
Any update on comments and plan for moving forward? I have one project depends on this PR. :-} 

thanks! 

Jeff.

**Comment by saisandeep** [comment]
> HI Saisandeep,
> 
> @saisandeep
> Any update on comments and plan for moving forward? I have one project depends on this PR. :-}
> 
> thanks!
> 
> Jeff.

Ok, will put this as a priority

**Comment by saisandeep** [comment]
@jeffhuang26 please review

**Review by jeffhuang26 — APPROVED** [review]

**Comment by jeffhuang26** [comment]
Any plan on merging this PR?

**Review by harshach — APPROVED** [review]
